packages/shared/ReactComponentStackFrame.js

```diff
@@ -19,6 +21,144 @@ import {
+  componentFrameCache = new PossiblyWeakMap();
+}
+
+export function describeNativeComponentFrame(
+  fn: Function,
+  construct: boolean,
+): string {
+  // If something asked for a stack inside a fake render, it should get ignored.
+  if (!fn || reentry) {
+    return '';
+  }
+
+  if (__DEV__) {
+    const frame = componentFrameCache.get(fn);
+    if (frame !== undefined) {
+      return frame;
+    }
+  }
+
+  const control = Error();
+
+  reentry = true;
+  let previousDispatcher;
+  if (__DEV__) {
+    previousDispatcher = ReactCurrentDispatcher.current;
+    // Set the dispatcher in DEV because this might be call in the render function
+    // for warnings.
+    ReactCurrentDispatcher.current = null;
+    disableLogs();
+  }
+  try {
+    // This should throw.
+    if (construct) {
+      // Something should be setting the props in the constructor.
+      const Fake = function() {};
+      // $FlowFixMe
+      Object.defineProperty(Fake.prototype, 'props', {
+        set: function() {
+          // We use a throwing setter instead of frozen or non-writable props
+          // because that won't throw in a non-strict mode function.
+          throw Error();
+        },
+      });
+      if (typeof Reflect === 'object' && Reflect.construct) {
+        Reflect.construct(fn, [], Fake);
+      } else {
+        fn.call(new Fake());
+      }
+    } else {
+      fn();
+    }
+  } catch (sample) {
+    // This is inlined manually because closure doesn't do it for us.
+    if (sample && typeof sample.stack === 'string') {
+      // This extracts the first frame from the sample that isn't also in the control.
+      // Skipping one frame that we assume is the frame that calls the two.
+      const sampleLines = sample.stack.split('\n');
+      const controlLines = control.stack.split('\n');
+      let s = sampleLines.length - 1;
+      let c = controlLines.length - 1;
+      while (s >= 1 && c >= 0 && sampleLines[s] !== controlLines[c]) {
+        // We expect at least one stack frame to be shared.
+        // Typically this will be the root most one. However, stack frames may be
+        // cut off due to maximum stack limits. In this case, one maybe cut off
+        // earlier than the other. We assume that the sample is longer or the same
+        // and there for cut off earlier. So we should find the root most frame in
+        // the sample somewhere in the control.
+        c--;
+      }
+      for (; s >= 1 && c >= 0; s--, c--) {
+        // Next we find the first one that isn't the same which should be the
+        // frame that called our sample function.
+        if (sampleLines[s] !== controlLines[c]) {
+          // Return the line we found.
+          // V8 adds a "new" prefix for native classes. Let's remove it to make it prettier.
+          const frame = '\n' + sampleLines[s - 1].replace(' at new ', ' at ');
+          if (__DEV__) {
+            if (typeof fn === 'function') {
```

Comment:
Why is this `typeof` check necessary?

Reply:
Because it throws in WeakMap if the wrong type sneaks in. That kills the stack which is not good if we're warning because some wrong type. We have tests that cover that.

Reply:
Gotcha. It looked odd to have `fn: Function` followed by a `typeof` check.